### PR TITLE
Optimizations for ImmutableArray

### PR DESCRIPTION
**Pull request by James-Ko**
This is a total of four commits so far to ``ImmutableArray`1.cs``.

 - Highlight: Wrapper-returning methods that used to return a `new ImmutableArray<T>(self.array)` now return `self`. ImmutableArray is a struct, overrides Equals, and is immutable, which makes 2 different ImmutableArrays pointing to the same array very very hard to distinguish from each other.

 - `InsertRange` now enumerates only once over the collection when inserting it into the new array. This also makes methods such as `InsertRange(int, ImmutableArray)` faster because that just inserts the inner array into this method, and when `ToArray()` detects that the enumerable is an array is just does a copy operation and never calls `foreach` at all.

(The downside to this is that we shouldn't even have to call `ToArray()` if the enumerable is already an `ICollection<T>`, so there should be some kind of overload of `InsertRange` for that.)

 - `ImmutableArray.Remove` and `ImmutableArray.Replace` now return `this` immediately if `this` is empty.

 - `Add` checks for empty, then calls `Insert`, which also checks for empty. Removed the check in `Add`.

 - Removed the lazy initialization of `List<int>` in `RemoveAll`: the default `List` ctor is very cheap (sets underlying array to an empty array), so it is definitely not worth the null check every iteration of the loop.

### CODE REVIEW

File: src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray`1.cs [hunk 1]
```diff
@@ -548,22 +548,17 @@ public ImmutableArray<T> InsertRange(int index, IEnumerable<T> items)
             {
                 return ImmutableArray.CreateRange(items);
             }
-
-            int count = ImmutableExtensions.GetCount(ref items);
-            if (count == 0)
+            
+            T[] other = items.ToArray();
```
Comment (sharwell) [inline src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray`1.cs]: :exclamation: I'm not sure this is better. It only iterates once, but it allocates a new intermediate array. When `items` implements `ICollection<T>`, I don't see how this is necessarily better.

File: src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray`1.cs [hunk 2]
```diff
@@ -691,6 +681,10 @@ public ImmutableArray<T> Replace(T oldValue, T newValue)
         public ImmutableArray<T> Replace(T oldValue, T newValue, IEqualityComparer<T> equalityComparer)
         {
             var self = this;
+            
+            if (self.IsEmpty)
```
Comment (sharwell) [inline src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray`1.cs]: :exclamation: This changes the semantics of the method. Previously, if `self` is empty the method always threw an exception.

File: src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray`1.cs [hunk 3]
```diff
@@ -875,25 +876,18 @@ public ImmutableArray<T> RemoveAll(Predicate<T> match)
             Requires.NotNull(match, "match");
 
             if (self.IsEmpty)
-            {
-                return new ImmutableArray<T>(self.array);
-            }
+                return self;
 
-            List<int> removeIndexes = null;
+            List<int> removeIndexes = new List<int>();
```
Comment (sharwell) [inline src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray`1.cs]: :exclamation: I'm not convinced this will be faster than the previous lazy initialization.

**Comment by sharwell** [comment]
I have a few concerns which are noted in comments in the diff. However, the wrapper elimination should be a small but easy win; I recommend separating those changes into their own pull request as I expect they would be readily merged.

File: src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray`1.cs (hunk 2 above)
Comment (James-Ko) [inline src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray`1.cs]: True, I'll roll this change back for `Replace`.

File: src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray`1.cs (hunk 1 above)
Comment (James-Ko) [inline src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray`1.cs]: @sharwell I have fixed this in a newer commit, at the cost of expanding the code by about 10 lines in a check for `ICollection<T>`. There is no longer a buffer between `items` and the dest array.

File: src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray`1.cs (hunk 3 above)
Comment (James-Ko) [inline src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray`1.cs]: @sharwell Why not? The [`List<T>` ctor](https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/Collections/Generic/List.cs#L49-L51) is extremely cheap so there's no benefit to lazy initialization. We also avoid a null check for every match, and make the code more readable.

File: src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray`1.cs [hunk 4]
```diff
@@ -545,26 +545,34 @@ public ImmutableArray<T> InsertRange(int index, IEnumerable<T> items)
...
             {
-                return self;
+                ICollection<T> other = (ICollection<T>)items;
+                
+                if (other.Count == 0)
+                    return self;
+                
+                T[] dest = new T[self.Length + other.Count];
+                Array.Copy(self.array, 0, dest, 0, index);
+                other.CopyTo(dest, index);
```
Comment (sharwell) [inline src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray`1.cs]: :exclamation: There is potential ability for an implementation of `CopyTo` to manipulate elements of the array which it shouldn't. You can solve this by moving the call to `other.CopyTo` above the two calls to `Array.Copy`, ensuring that even if the call to `CopyTo` results in out-of-bounds array writes, the elements before and after the inserted range are fully protected against modification.

File: src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray`1.cs (hunk 3 above)
Comment (sharwell) [inline src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray`1.cs]: In the event no item is removed, this code allocates a `List<int>` instance which is never used. I believe this change deserves a much closer look during the code review, which is why I pointed it out.

File: src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray`1.cs (hunk 4 above)
Comment (James-Ko) [inline src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray`1.cs]: @sharwell Fixed. That's a pretty sharp insight.

File: src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray`1.cs (hunk 3 above)
Comment (James-Ko) [inline src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray`1.cs]: I have rewritten the method to both avoid the null check and lazily instantiate the `List<int>` in the new commit.

**Comment by mikedn** [comment]
> Highlight: Wrapper-returning methods that used to return a  new ImmutableArray<T>(self.array)  now return  self . ImmutableArray is a struct, overrides Equals, and is immutable, which makes 2 different ImmutableArrays pointing to the same array very very hard to distinguish from each other.

It's not very hard, it is impossible. The value the caller gets is always a copy of whatever value you pass to return, be it `self` or `new ImmutableArray<T>(self.array)`. In terms of generated code this particular change has no effect, fortunately the JIT manages to correctly handle structs that contain only one reference field. Not to say that the change is wrong and should not be done, it's just that it doesn't do what you think it does.

File: src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray`1.cs [hunk 5]
```diff
@@ -545,26 +545,34 @@ public ImmutableArray<T> InsertRange(int index, IEnumerable<T> items)
...
+                T[] other = items.ToArray();
+                
+                if (other.Length == 0)
+                    return self;
+    
+                T[] tmp = new T[self.Length + other.Length];
+                Array.Copy(self.array, 0, tmp, 0, index);
+                Array.Copy(other, 0, tmp, index, other.Length);
+                Array.Copy(self.array, index, tmp, index + other.Length, self.Length - index);
+                return new ImmutableArray<T>(tmp);
```
Comment (stephentoub) [inline src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray`1.cs]: This change impacts both behavior and performance, and not necessarily in a positive way.  For example, the previous implementation was using GetCount, which is able to get a count from an ```ICollection```, an ```ICollection<T>```, or an ```IReadOnlyCollection<T>```, whereas the new version is only specializing ```ICollection<T>```.  There's too much changing as part of this for me to feel confident that it's a net benefit.  Do you have performance tests that demonstrate it is across a range of input collections and collection types?  If so, can you share?  I suggest this be reverted for now.

File: src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray`1.cs [hunk 6]
```diff
@@ -602,11 +610,6 @@ public ImmutableArray<T> InsertRange(int index, ImmutableArray<T> items)
         public ImmutableArray<T> Add(T item)
         {
             var self = this;
-            if (self.Length == 0)
-            {
-                return ImmutableArray.Create(item);
-            }
-
```
Comment (stephentoub) [inline src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray`1.cs]: It's true that Insert does a length check, but that doesn't necessarily mean this should be eliminated.  For example, if Add can typically be inlined, Insert can't, and the case of adding to an empty list is common (which I believe it is), then this could actually be hurting performance in a common case because it forces the calling of a non-inlined method.  We shouldn't be making/taking changes like this without a solid set of perf measurements across a range of scenarios to demonstrate that it's actually a win.

File: src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray`1.cs [hunk 7]
```diff
@@ -726,9 +730,13 @@ public ImmutableArray<T> Remove(T item, IEqualityComparer<T> equalityComparer)
         {
             var self = this;
             self.ThrowNullRefIfNotInitialized();
+            
+            if (self.IsEmpty)
+                return self;
+
```
Comment (stephentoub) [inline src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray`1.cs]: Is it common to Remove from an empty list?  Upfront checks like this that optimize for the empty case are great if we expect it to be empty, but if it's more likely for it not to be empty, then this actualy penalizes the common case.

File: src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray`1.cs [hunk 8]
```diff
@@ -797,6 +805,9 @@ public ImmutableArray<T> RemoveRange(IEnumerable<T> items, IEqualityComparer<T>
             self.ThrowNullRefIfNotInitialized();
             Requires.NotNull(items, "items");
             Requires.NotNull(equalityComparer, "equalityComparer");
+            
+            if (self.IsEmpty)
+                return self;
```
Comment (stephentoub) [inline src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray`1.cs]: Ditto.

File: src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray`1.cs [hunk 9]
```diff
@@ -839,11 +850,11 @@ public ImmutableArray<T> RemoveRange(ImmutableArray<T> items)
         public ImmutableArray<T> RemoveRange(ImmutableArray<T> items, IEqualityComparer<T> equalityComparer)
         {
             var self = this;
+            self.ThrowNullRefIfNotInitialized();
```
Comment (stephentoub) [inline src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray`1.cs]: This is changing behavior.  Why did you make this change?

File: src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray`1.cs [hunk 10]
```diff
@@ -875,23 +886,25 @@ public ImmutableArray<T> RemoveAll(Predicate<T> match)
             Requires.NotNull(match, "match");
 
             if (self.IsEmpty)
-            {
-                return new ImmutableArray<T>(self.array);
-            }
+                return self;
```
Comment (stephentoub) [inline src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray`1.cs]: Even if we change this to ```return self;``` instead of ```return new ImmutableArray<T>(self.array);```, we should still keep the braces that keep the formatting consistent as much as possible with the rest of the project.

File: src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray`1.cs [hunk 11]
```diff
@@ -875,23 +886,25 @@ public ImmutableArray<T> RemoveAll(Predicate<T> match)
...
+                    removeIndexes = new List<int>();
                     removeIndexes.Add(i);
+                    break;
                 }
             }
+            for (; i < self.array.Length; i++)
+            {
+                if (match(self.array[i]))
+                    removeIndexes.Add(i);
+            }
```
Comment (stephentoub) [inline src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray`1.cs]: Could you share your performance test that demonstrates this change helps?  I understand it in theory.  In practice it makes the code more complicated and duplicates code, so there should be a measurable win to justify it.

File: src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray`1.cs (hunk 9 above)
Comment (sharwell) [inline src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray`1.cs]: :memo: The only case where this changes behavior is when *both* `this` and `items` are default instances of `ImmutableArray<T>` - the change is due to the reversal in the order of checking. Previously this threw an `ArgumentNullException` for this case, but now it throws a `NullReferenceException`.

:thought_balloon: The new behavior appears to be more consistent with other methods in this type.

File: src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray`1.cs (hunk 8 above)
Comment (stephentoub) [inline src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray`1.cs]: This change is also the cause of the test failure, I believe, as it also changes behavior, and there's a test that catches that.

**Comment by stephentoub** [comment]
cc: @AArnott

File: src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray`1.cs (hunk 9 above)
Comment (James-Ko) [inline src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray`1.cs]: @stephentoub This was the only case in the file where the init check was not at the top of the method, so I thought it may have been a mistake on the part of the original writer.

Comment (stephentoub) [inline src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray`1.cs]: @James-Ko, it may very well have been a mistake, and it may very well make things more consistent with the rest of the system... I can appreciate both of those things.  It's just that at this point the library has already shipped, so we need to be cognizant of any changes in behavior and either avoid them or at least be very thoughtful about when and why they're made.  Such changes should not be made as part of a PR focused on performance improvements.  Thanks for understanding.

**Comment by AArnott** [comment]
Thanks for the CC, @stephentoub. I am very interested in reviewing before this is merged. But I can't review right now. Perhaps by Monday?

**Comment by stephentoub** [comment]
@AArnott, have you had a chance to look at this? I left a bunch of comments highlighting concerns, and I'm wondering if there are pieces of this PR you think should be kept.

**Comment by AArnott** [comment]
Not yet, but I should be able to find some time today to.

File: src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray`1.cs (hunk 9 above)
Comment (AArnott) [inline src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray`1.cs]: +1 to both of @stephentoub's arguments: yes possibly this is a good change, and yes it should be its own PR.

File: src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray`1.cs (hunk 10 above)
Comment (AArnott) [inline src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray`1.cs]: There are a few of these changes in this file (returning an existing instance of the struct rather than creating a new one). Is this a micro-optimization where we avoid creating/copying a struct one additional time? If so, I don't object. This makes it slightly easier to read and (I guess) slightly faster. But that probably needs a new, dedicated pull request to get away from several of the more questionable changes in this PR.

File: src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray`1.cs (hunk 11 above)
Comment (AArnott) [inline src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray`1.cs]: +1

Comment (AArnott) [inline src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray`1.cs]: +1
I'd also ask that comments explaining the code be added since it's not as obvious why it's written the way you've done it.

**Comment by AArnott** [comment]
@James-Ko Thanks for your time in contributing this. I imagine you took quite a bit of time to look for these micro-optimizations. I agree with many of @stephentoub's comments however. I recommend we close this PR and if you'd like to send a few more focused PRs (each with one kind of optimization applied throughout the class, and without unnecessary formatting changes) for those ideas that sounded good but were lost amidst the noise of the other changes, we may be able to entertain those.
Where we asked for perf numbers to justify specific changes, I think we should get those before taking those kinds of changes.

**Comment by James-Ko** [comment]
@AArnott Sure; I've been busy so I didn't have the time to separate the changes like @sharwell suggested, but I'll close this and do that right now.
